internal/system/agent/services.go

```diff
@@ -34,6 +37,193 @@ const (
 	RESTART = "restart"
 )
 
+func InvokeMetrics(services []string, ctx context.Context) (MetricsRespMap, error) {
+
+	m := MetricsRespMap{}
+	m.Metrics = map[string]interface{}{}
+
+	// Loop through requested actions, along with (any) respectively-supplied parameters.
+	for _, service := range services {
+		LoggingClient.Info("invoke metrics")
+
+		if Configuration.MetricsMechanism == "direct-service" {
+			es := ExecuteService{}
+			out, err := es.Metrics(ctx, service)
+			if err != nil {
+				LoggingClient.Error("error fetching metrics")
+				return m, err
+			} else {
+				m.Metrics[service] = ProcessResponse(string(out))
+			}
+		} else if Configuration.MetricsMechanism == "executor" {
+			var result []byte
+			ea := ExecuteApp{}
+			out, err := ea.Metrics(ctx, service)
+			if err.Error() != "exit status 1" {
+				LoggingClient.Error("error fetching metrics")
+				return m, err
+			} else {
+				result, _ = processOutput(out, service)
+				m.Metrics[service] = ProcessResponse(string(result))
+			}
+		} else if Configuration.MetricsMechanism == "custom" {
+			err := fmt.Errorf("the requested custom executor (e.g. snap) has not been integrated")
+			LoggingClient.Error(err.Error())
+			m.Metrics[service] = fmt.Sprintf(err.Error())
+		} else {
+			err := fmt.Errorf("the requested metrics mechanism is not supported")
+			LoggingClient.Error(err.Error())
+		}
+	}
+	return m, nil
+}
+
+func processOutput(bytes []byte, service string) ([]byte, error) {
+
+	var relevantLines []string
+	s := string(bytes)
+	var fields [][]string
+	var populated = map[string]Stats{}
+
+	lines, err := stringToLines(s)
+	if err != nil {
+		LoggingClient.Error(err.Error())
+		return nil, err
+	}
+	relevantLines = findRelevantLines(lines)
+	fields = tabulateResult(relevantLines)
+	populated = populateMap(fields)
+	relevantStats := selectRelevantStats(populated, service)
+	response, err := assembleStats(relevantStats)
+	if err != nil {
+		LoggingClient.Error(err.Error())
+		return nil, err
+	}
+	return response, nil
+}
+
+func assembleStats(stats map[string]Stats) ([]byte, error) {
+
+	const (
+		CpuPercent       = "CPU %"
+		MemUsageAndLimit = "MEM USAGE / LIMIT"
+		MemPercent       = "MEM %"
+		Net_IO           = "NET I/O"
+		Block_IO         = "BLOCK I/O"
+		Pids             = "PIDS"
+	)
+
+	data := make(map[string]string)
+
+	for _, value := range stats {
+		data[CpuPercent] = value.CpuPercent
+		data[MemUsageAndLimit] = value.MemUsageAndLimit
+		data[MemPercent] = value.MemPercent
+		data[Net_IO] = value.Net_IO
+		data[Block_IO] = value.Block_IO
+		data[Pids] = value.Pids
+	}
+
+	assembled, err := json.Marshal(data)
+	if err != nil {
+		LoggingClient.Error(err.Error())
+		return nil, err
+	}
+	return assembled, err
+}
+
+func selectRelevantStats(stats map[string]Stats, service string) map[string]Stats {
```

Comment:
I don't quite follow the purpose of this function - it seems like this function is just getting the specified service key from the map? I think this could be much more efficiently implemented by just accessing the map directly.